wallet: Migrate legacy wallets to descriptor wallets

achow101 opened the pull request:
This PR adds a new `migratewallet` RPC which migrates a legacy wallet to a descriptor wallet. Migrated wallets will need a new backup. If a wallet has watchonly stuff in it, a new watchonly descriptor wallet will be created containing those watchonly things. The related transactions, labels, and descriptors for those watchonly things will be removed from the original wallet. Migrated wallets will not have any of the legacy things be available for fetching from `getnewaddress` or `getrawchangeaddress`. Wallets that have private keys enabled will have newly generated descriptors. Wallets with private keys disabled will not have any active `ScriptPubKeyMan`s.

For the basic HD wallet case of just generated keys, in addition to the standard descriptor wallet descriptors using the master key derived from the pre-existing hd seed, the migration will also create 3 descriptors for each HD chain in: a ranged combo external, a ranged combo internal, and a single key combo for the seed (the seed is a valid key that we can receive coins at!). The migrated wallet will then have newly generated descriptors as the active `ScriptPubKeyMan`s. This is equivalent to creating a new descriptor wallet and importing the 3 descriptors for each HD chain. For wallets containing non-HD keys, each key will have its own combo descriptor.

There are also tests.

achow101 committed:
scriptpubkeyman: Implement GetScriptPubKeys in Legacy

achow101 committed:
Implement LegacyScriptPubKeyMan::MigrateToDescriptor

achow101 committed:
Implement MigrateLegacyToDescriptor

achow101 committed:
Add migratewallet RPC

achow101 committed:
descriptors: addr() and raw() should return false for ToPrivateString

They don't have any private data and they can't be nested so they
should return false for ToPrivateString.

achow101 committed:
Test migratewallet

Co-Authored-By: furszy <matiasfurszyfer@protonmail.com>

hebasto commented:
@achow101 
It seems this PR has [problems](https://travis-ci.org/github/bitcoin/bitcoin/jobs/712335127) with linters.

luke-jr commented:
Why doesn't this extend `upgradewallet`?

achow101 commented:
> Why doesn't this extend `upgradewallet`?

It didn't feel like it fit `upgradewallet` as I see that more for changing the wallet version number and using legacy wallet features. But this is changing a legacy wallet to something completely new.

This is also pretty invasive and fundamentally changes how the wallet is working, so I wanted to keep it separate from something that people might still want to use on legacy wallets.

laanwj commented:
I think we should only add this functionality after `sqlite` wallets? Otherwise you'd keep migrating.

achow101 commented:
> I think we should only add this functionality after `sqlite` wallets? Otherwise you'd keep migrating.

I think there's a question of whether we want to keep sqlite separate from descriptors. We might want to allow legacy wallets to migrate to sqlite without migrating to descriptors since they are orthogonal.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4113,6 +4113,33 @@ static UniValue upgradewallet(const JSONRPCRequest& request)
     return error.original;
 }
 
+static UniValue migratewallet(const JSONRPCRequest& request)
```
nit: might be good to use the non-deprecated constructor to avoid having to change this again. See e.g. #19528

achow101 replied:
Getting a build error when I try to do this:

```wallet/rpcwallet.cpp:4280:5: error: no matching constructor for initialization of 'const CRPCCommand'
    { "wallet",             "migratewallet",                    &migratewallet,                 {} },
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./rpc/server.h:120:5: note: candidate constructor not viable: no known conversion from 'RPCHelpMan (*)(const JSONRPCRequest &)' to 'rpcfn_type' (aka 'UniValue (*)(const JSONRPCRequest &)') for 3rd argument
    CRPCCommand(const char* category, const char* name, rpcfn_type fn, std::initializer_list<const char*> args)
    ^
./rpc/server.h:107:5: note: candidate constructor not viable: no known conversion from 'RPCHelpMan (*)(const JSONRPCRequest &)' to 'RpcMethodFnType' (aka 'RPCHelpMan (*)()') for 3rd argument
    CRPCCommand(std::string category, std::string name_in, RpcMethodFnType fn, std::vector<std::string> args_in)
    ^
./rpc/server.h:100:5: note: candidate constructor not viable: requires 5 arguments, but 4 were provided
    CRPCCommand(std::string category, std::string name, Actor actor, std::vector<std::string> args, intptr_t unique_id)
    ^
./rpc/server.h:91:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 4 were provided
class CRPCCommand
      ^
./rpc/server.h:91:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 4 were provided
1 error generated.
```

MarcoFalke replied:
Not sure why. This diff compiles for me:

```diff
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index cc89c4992b..1361955a4b 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -4171,9 +4171,9 @@ static UniValue upgradewallet(const JSONRPCRequest& request)
     return error.original;
 }
 
-static UniValue migratewallet(const JSONRPCRequest& request)
+static RPCHelpMan migratewallet()
 {
-    RPCHelpMan{"migratewallet",
+    return RPCHelpMan{"migratewallet",
         "\nMigrate the wallet to a descriptor wallet.\n"
         "A new wallet backup will need to be made.",
         {},
@@ -4181,9 +4181,9 @@ static UniValue migratewallet(const JSONRPCRequest& request)
         RPCExamples{
             HelpExampleCli("migratewallet", "")
             + HelpExampleRpc("migratewallet", "")
-        }
-    }.Check(request);
-
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
     std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
     if (!wallet) return NullUniValue;
     CWallet* const pwallet = wallet.get();
@@ -4196,6 +4196,8 @@ static UniValue migratewallet(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_WALLET_ERROR, error.original);
     }
     return error.original;
+},
+    };
 }
 
 RPCHelpMan abortrescan();

achow101 replied:
Ah, I seem to have made a mistake somewhere.

maflcko reviewed (commented):
Left a style-nit. Feel free to ignore.

meshcollider commented:
Do you know what the performance of a wallet would be like, in the case of a non-HD wallet being migrated?

achow101 commented:
> Do you know what the performance of a wallet would be like, in the case of a non-HD wallet being migrated?

Probably worse.

meshcollider commented:
> Probably worse.

Almost certainly, that's why I'm asking :) It would be nice to know how much worse it would be

Sjors commented:
Concept ACK, but I prefer this to be in the wallet tool.

achow101 commented:
> but I prefer this to be in the wallet tool.

I would prefer this to be accessible to most users so that more people can move to using descriptor wallets. For most wallets, this should be painless and not result in a horribly bloated wallet.

Sjors commented:
Possibly spurious Travis [error](https://travis-ci.org/github/bitcoin/bitcoin/jobs/722952265#L2561):
```
020-09-01T05:18:47.709000Z TestFramework (INFO): Initializing test directory /Users/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20200901_051608/wallet_basic_178
...
25662020-09-01T05:19:50.052000Z TestFramework (INFO): check -reindex
25672020-09-01T05:20:03.516000Z TestFramework (ERROR): Assertion failed
2568Traceback (most recent call last):
...
2575test_framework.authproxy.JSONRPCException: Amount out of range (-3)
2576
2577During handling of the above exception, another exception occurred:
2578
2579Traceback (most recent call last):
2580  File "/Users/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin16/test/functional/test_framework/test_framework.py", line 118, in main
2581    self.run_test()
2582  File "/Users/travis/build/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin16/test/functional/wallet_basic.py", line 601, in run_test
2583    assert_raises_rpc_error(-6, "Transaction has too long of a mempool chain", self.nodes[0].sendtoaddress, sending_addr, node0_balance - Decimal('0.01'))
```

Sjors commented:
The wallet tool is just as available as the RPC. We can print the necessary incantation in the help if needed. It does make sense to have an upgrade button in the GUI though.

achow101 commented:
> The wallet tool is just as available as the RPC.

Not really. The RPC has a dedicated window within the GUI. The wallet tool requires actually using a terminal.

jonatack commented:
Concept ACK, a few quick thoughts before reviewing the code tomorrow:
- debug build clean and local tests green
- the help in `bitcoin-cli help migratewallet` is pretty sparse? -- could benefit from a lot more helpful info to explain what/why/how to users, de-risk wallet migrations a bit and encourage users to move to descriptor wallets
- agree with @meshcollider that performance info would be good here in the PR and in the doc (`descriptors.md` or a new one?) and/or rpc help 
- test is missing logging

achow101 commented:
Hmm, I think the scriptPubKey set we are generating here is incorrect. I'm going to work on a slightly separate thing for LegacyScriptPubKeyMan that ensures we have the correct scriptPubKey set.

achow101 commented:
Fixed the IsMine issue and added some tests for that. Also rebased onto master as there was a hidden merge conflict.

>     * the help in `bitcoin-cli help migratewallet` is pretty sparse? -- could benefit from a lot more helpful info to explain what/why/how to users, de-risk wallet migrations a bit and encourage users to move to descriptor wallets

Any suggestions?

>     * agree with @meshcollider that performance info would be good here in the PR and in the doc (`descriptors.md` or a new one?) and/or rpc help

I'll do that eventually. After I figure out how.

>     * test is missing logging

Added some logging

jonatack commented:
Thanks @achow101, I need to review and test this.

adamjonas commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,280 @@
+
+import os
+
+from test_framework.key import ECKey
```
The linter is complaining this going unused.

achow101 replied:
Fixed.

MarcoFalke commented:
Needs rebase

MarcoFalke commented:
```
SUMMARY: ThreadSanitizer: data race /usr/lib/llvm-10/bin/../include/c++/v1/memory:2616:20 in std::__1::unique_ptr<WalletDatabase, std::__1::default_delete<WalletDatabase> >::reset(WalletDatabase*)
```

https://cirrus-ci.com/task/5621550420328448?command=ci#L4781

laanwj commented on src/wallet/wallet.cpp:
```diff
@@ -4567,3 +4567,287 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+    fs::path db_dir = db_path.branch_path();
+    std::string db_filename = db_path.leaf().string();
+    m_database->Close();
+    fs::remove(db_path);
```
This looks scary to me. Let's not delete the old wallet, but rename it. You know, just in case something with the migration didn't go well it's good to have a backup.

achow101 replied:
~20 lines earlier we make a backup of the wallet. So there is still a backup and we're just removing the original after it is backed up.

laanwj replied:
Okay, sounds good to me then.

laanwj commented:
Adding to the 23.0 milestone, I think it would be good to have optional migration in as soon as possible.

prayank23 commented:
Concept ACK

katesalazar commented:
What's the minimum wallet version this tool will support?

MarcoFalke commented:
Needs rebase

achow101 commented:
Rebased

sipa commented:
> and a single key combo for the seed (the seed is a valid key that we can receive coins at!)

🤦

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4839,6 +4839,37 @@ static RPCHelpMan walletdisplayaddress()
+    WalletContext& context = EnsureWalletContext(request.context);
+
+    bilingual_str error;
+    std::vector<bilingual_str> warnings;
```
is this ignored?

achow101 replied:
Removed, as well as in the `Migrate*` functions that aren't actually using `warnings`.

MarcoFalke commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1613,6 +1613,351 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
+
+    MigrationData out;
+
+    std::unordered_set<CScript, SaltedSipHasher> spks;
+    auto spks_temp = GetScriptPubKeys();
+    spks.insert(spks_temp.begin(), spks_temp.end());
```
```suggestion
    auto spks{GetScriptPubKeys()};
```

Does this not compile?

achow101 replied:
Done

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -3335,3 +3335,328 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+        context.wallets.erase(i);
+    }
+    // Now do the database stuff
+    if (!wallet.MigrateToSQLite(error ,warnings)) return false;
```
clang-format?

achow101 replied:
No longer relevant.

maflcko reviewed (commented):
left some nits/questions

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -3335,3 +3335,330 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+            wallet.WalletLogPrintf("Making a new watchonly wallet containing the watched scripts\n");
+
+            DatabaseStatus status;
+            std::vector<bilingual_str> warnings;
```
well, now you are ignoring those?

achow101 replied:
I don't think that those warnings matter. There aren't any warnings for newly created wallets. These were being ignored originally anyways.

MarcoFalke commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,288 @@
+        assert_equal(basic1.getwalletinfo()["descriptors"], False)
+
+        for i in range(0, 10):
+            default.sendtoaddress(basic1.getnewaddress(), 1)
```
test_framework.authproxy.JSONRPCException: Requested wallet does not exist or is not loaded (-18)


https://cirrus-ci.com/task/6547488835371008?logs=ci#L4027

amovfx reviewed (commented):
Went through this for the bitcoin pr review club

michaelfolkson commented:
Concept ACK, Approach ACK. 

Definitely need a migration tool from legacy and/or BDB to descriptor and sqlite. I haven't tested yet but the instructions I intend to follow are:

- Create signet legacy-bdb wallet
- Generate multiple signet addresses
- Sends funds to a particular signet address
- Use the migratewallet RPC in this PR
- Check the funds are now stored at the equivalent descriptor, the other generated addresses are there too (the migration generally worked fine)

Sjors commented:
I'm tempted to suggest that we shouldn't touch the original wallet, especially with the way this handles watch-only addresses. Maybe it's better to create two new wallets (if there's any watch-only address), copy the relevant transactions over and then unload the original wallet. Perhaps we can even add a flag to the original wallet to disable new address generation?

achow101 commented:
> I'm tempted to suggest that we shouldn't touch the original wallet, especially with the way this handles watch-only addresses. Maybe it's better to create two new wallets (if there's any watch-only address), copy the relevant transactions over and then unload the original wallet. Perhaps we can even add a flag to the original wallet to disable new address generation?

I think it might be confusing to have the original wallet under the original name, and then the migrated wallets with new names. Particularly if users have specified their wallet(s) in the bitcoin.conf file, as then the original would still always be opened. I suppose with settings.json we can at least ensure that the migrated ones are also always opened.

Sjors commented:
Maybe rename the old one to `…-archive-DATE`?

achow101 commented:
> Maybe rename the old one to `…-archive-DATE`?

We already make a backup of it of the form `name-timestamp.legacy.bak`

Sjors commented:
True. I suppose the upgrade is only potentially confusing for users with a mix of spendable and watch-only addresses. So perhaps there could be two seperate flows:
1. Spendable OR watch-only wallets: as implemented
2. Mixed wallet: keep the .legacy.bak wallet open and loaded, create the two wallets as you do now, but move the transactions to the correct wallet. Have a popup explain the split, and tell user they can should close .legacy.bak wallet when done.

In this case my suggestion to add a flag to prevent new address generation wouldn't work, because we probably shouldn't touch the backup at all. Using the same master seed in the new wallet should work though?

achow101 commented:
> 2. Mixed wallet: keep the .legacy.bak wallet open and loaded,

I don't think it's necessary to open the backup and then require the user to close it.

> Using the same master seed in the new wallet should work though?

That can be done.

achow101 committed:
wallet: Refactor SetupDescSPKMs to take CExtKey

Refactors SetupDescSPKMs so that the DescSPKM loops are in their own
function. This allows us to call it later during migration with a key
that was already generated.

Sjors commented:
> I don't think it's necessary to open the backup and then require the user to close it.

Another idea could be to refuse the migration if the wallet contains a combination of private key and watch-only stuff. Or to require that user specifies what the behaviour should be (keeping them in the same wallet or moving them to a fresh watch-only wallet).

achow101 commented:
> Another idea could be to refuse the migration if the wallet contains a combination of private key and watch-only stuff. Or to require that user specifies what the behaviour should be (keeping them in the same wallet or moving them to a fresh watch-only wallet).

I think anything that prevents the user from migrating or requires the user to make decisions or jump through hoops to do the migration is just going to result in people not migrating their wallets. We need people to migrate their wallets, and for the process to go as smoothly as possible, so that we can get rid of the legacy wallet.

achow101 committed:
doc: Release notes and other docs for migration

achow101 commented:
I've added an experimental warning on the RPC and some release notes describing what migratewallet will do and some additional warnings and instructions for users.

furszy commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1707,6 +1707,241 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+
+    std::unordered_set<CScript, SaltedSipHasher> spks{GetScriptPubKeys()};
+
+    // Get all key ids
+    std::set<CKeyID> keyids;
+    for (const auto& key_pair : mapKeys) {
+        keyids.insert(key_pair.first);
+    }
+    for (const auto& key_pair : mapCryptedKeys) {
+        keyids.insert(key_pair.first);
+    }
+
+    // Get key metadata and figure out which keys don't have a seed
+    // Note that we do not ignore the seeds themselves because they are considered IsMine!
+    for (auto keyid_it = keyids.begin(); keyid_it != keyids.end();) {
+        const CKeyID& keyid = *keyid_it;
+        const auto& it = mapKeyMetadata.find(keyid);
+        if (it != mapKeyMetadata.end()) {
+            const CKeyMetadata& meta = it->second;
+            if (meta.hdKeypath == "s" || meta.hdKeypath == "m") {
+                keyid_it++;
+                continue;
+            }
+            if (m_hd_chain.seed_id == meta.hd_seed_id || m_inactive_hd_chains.count(meta.hd_seed_id) > 0) {
+                keyid_it = keyids.erase(keyid_it);
+                continue;
+            }
+        }
+        keyid_it++;
+    }
```
Guess that you are looping first through `mapKeys` and `mapCryptedKeys` because in non-HD wallets we were not storing an entry inside `mapKeyMetadata` for each key?

Because otherwise, if we always have an entry inside `mapKeyMetadata` for each key, we could just loop through `mapKeyMetadata` to generate each single key combo descriptor directly.

achow101 replied:
There are some situations where a key record can exist without a corresponding metadata record. So it's safer to go through all of the keys and pull up their metadata than just iterating `mapKeyMetadata`

furszy commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1947,6 +1947,69 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
+bool LegacyScriptPubKeyMan::DeleteRecords(bilingual_str& error)
+{
+    LOCK(cs_KeyStore);
+    WalletBatch batch(m_storage.GetDatabase());
```
let's begin a db txn here and commit it at the end of the function.

otherwise if an error arises in the middle, we will end up with half of the information removed.

achow101 replied:
Good point.

furszy reviewed (commented):
had an issue during migration (unknown cause..) and half of the info got removed from disk while the other half remained there.

now every time that I start the wallet, I get "seed not found" which means that the key got removed but the hd chain seed id not.

So, left a comment to solve the non-recoverable state at least. We should either remove all the records in a single atomic write or not remove them at all and fail.

furszy reviewed (commented):
The PR description doesn't seems to be accurate to the current sources. It says:

> For the basic HD wallet case of just generated keys, the migration will create 3 descriptors for each HD chain: a ranged combo external, a ranged combo internal, and a single key combo for the seed (the seed is a valid key that we can receive coins at!)

And, for the basic case of an HD wallet, we are actually creating 11 descriptors on the migration process:

* BIP32 descriptors; form of "0'/0'/*" and "0'/1'/*" (2 descriptors)
* BIP44 descriptors; "44'/1'/0'/0/*" and "44'/1'/0'/1/*" (2 descriptors)
* BIP49 descriptors, P2SH(P2WPKH);  form of "49'/1'/0'/0/*" and "49'/1'/0'/1/*" (2 descriptors)
* BIP84 descriptors, P2WPKH; form of "84'/1'/0'/1/*" and "84'/1'/0'/1/*" (2 descriptors)
* BIP86 descriptors, P2TR; form of "86'/1'/0'/0/*" and "86'/1'/0'/1/*" (2 descriptors)
* A combo(PK) descriptor for the wallet master key. 

Then, while was checking what was going on, have expanded the test coverage of the basic HD wallet migration case with the 11 descriptors check and the following points:

1) Test case for a wallet with balance received on the seed.
   Receiving coins in different outputs created from the seed key (P2PKH, P2WPKH, P2SH(P2WPKH)) and verifying that after migration the wallet still has them.

2) Verify `getaddressinfo` remains unchanged after migration.
   (solvability, hdkeypath, ischange, hdmasterfingerprint)

3) Verify that the migrated wallet was flushed to disk by restarting the node and checking that balance/txes are still there.

If you like it, can cherry-pick it from: https://github.com/furszy/bitcoin/commit/e31779c3edca96a4a3f26ab554f7b9824ea0338b

The more coverage we can get over this, the better.
Will continue reviewing :).

achow101 commented:
> > For the basic HD wallet case of just generated keys, the migration will create 3 descriptors for each HD chain: a ranged combo external, a ranged combo internal, and a single key combo for the seed (the seed is a valid key that we can receive coins at!)
> 
> And, for the basic case of an HD wallet, we are actually creating 11 descriptors on the migration process:

It was meant to be read as "in addition to the usual descriptors". I've updated the description to clarify that.

achow101 commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,289 @@
+        self.skip_if_no_sqlite()
+        self.skip_if_no_bdb()
+
+    def assert_is_sqlite(self, wallet_name):
+        wallet_file_path = os.path.join(self.nodes[0].datadir, "regtest/wallets", wallet_name, self.wallet_data_filename)
+        with open(wallet_file_path, 'rb') as f:
+            file_magic = f.read(16)
+            assert_equal(file_magic, b'SQLite format 3\x00')
```
For these upgrade tests, I prefer to have checks external to the actual code as in-memory state could be different from on-disk state.

Sjors commented on doc/release-notes-19602.md:
```diff
@@ -0,0 +1,25 @@
+makes a best effort attempt to capture all of these things into Descriptor wallets. There may be
+unforseen configurations which result in some scripts being excluded. If a migration fails
+unexpectedly or otherwise misses any scripts, please create an issue on GitHub. A backup of the
+original wallet can be found in the wallet directory with the name `<name>-<timestamp>.legacy.bak`.
```
What should the user do if there's a .log file? Delete it?

Sjors replied:
(in the context of restoring a backup, since it's a bit unclear if the log file belongs to the backup, the current version or if it's just an artefact from before the backup was made, and safe to delete)

achow101 replied:
It's safe to delete. The backup uses our normal backup code which compacts everything into the db so log files are no longer needed.

Sjors commented:
@achow101 I sent you a signet wallet with some watch-only keys that didn't migrate correctly.

achow101 commented:
> @achow101 I sent you a signet wallet with some watch-only keys that didn't migrate correctly.

This should be fixed now. Also added a few test cases for that failure.

achow101 commented:
Latest push adds cleanup on failure.

darosior commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,36 @@ RPCHelpMan simulaterawtransaction()
+static RPCHelpMan migratewallet()
+{
+    return RPCHelpMan{"migratewallet",
+        "\nEXPERIMENTAL warning: This call may not work as expected and may be changed in future releases\n"
+        "\nMigrate the wallet to a descriptor wallet.\n"
+        "A new wallet backup will need to be made.",
```
Maybe this could mention that a backup will be created in the same manner as by using `backupwallet`, that can be restored using `restorewallet`?

achow101 replied:
Added some text about that.

darosior commented on doc/managing-wallets.md:
```diff
@@ -120,4 +120,29 @@ After that, `getwalletinfo` can be used to check if the wallet has been fully re
+additional wallets created after migration. In addition to a descriptor wallet of the same name,
+there may also be a wallet named `<name>_watchonly` and `<name>_solvables`. `<name>_watchonly`
+contains all of the watchonly scripts. `<name>_solvables` contains any scripts which the wallet
+knows the but is not watching the corresponding P2(W)SH scripts.
```
nit: typo "knows the but is not watching"

achow101 replied:
Fixed.

darosior commented on doc/managing-wallets.md:
```diff
@@ -120,4 +120,29 @@ After that, `getwalletinfo` can be used to check if the wallet has been fully re
\ No newline at end of file
+The restored wallet can also be loaded in the GUI via `File` ->`Open wallet`.
+
+## Migrating Legacy Wallets to Descriptor Wallets
```
Thanks for adding documentation here.

nit: instead of duplicating the text in the release notes, maybe we could link to this document? Also, we could link to this document from places were we say BDB is legacy (such as in build-UNIX.md for instance)?

achow101 replied:
Done

darosior reviewed (commented):
I unfortunately won't have time to review the code before feature freeze, but at least i successfully tested this with a mainnet watchonly legacy wallet (with a couple hundred transactions and a couple thousands entries in the address book) along with a legacy testnet wallets (with private keys) with a few hundreds transactions and a couple dozen entries in the address book.

Sjors commented:
It would be good to explicitly test wallets with coinbase transactions on `pk()` "addresses", since afaik that's what the built-in miner used.

achow101 commented:
> It would be good to explicitly test wallets with coinbase transactions on `pk()` "addresses", since afaik that's what the built-in miner used.

Done

w0xlt commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1658,6 +1658,360 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
+    }
+
+    // Handle HD keys by using the CHDChains
+    std::set<CHDChain> chains;
+    chains.insert(m_hd_chain);
+    for (const auto& chain_pair : m_inactive_hd_chains) {
+        chains.insert(chain_pair.second);
+    }
```
Could this be a vector, assuming `m_inactive_hd_chains` doesn't store repeated `CHDChain` items ?
This way, changing `src/wallet/walletdb.h` would not be necessary..

```suggestion
    std::vector<CHDChain> chains;
    chains.push_back(m_hd_chain);
    for (const auto& chain_pair : m_inactive_hd_chains) {
        chains.push_back(chain_pair.second);
    }
```

achow101 replied:
Done

w0xlt reviewed (commented):
Approach ACK

w0xlt commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1658,6 +1658,360 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
+    for (const CScript& script : setWatchOnly) {
+        if (!batch.EraseWatchOnly(script)) {
+            error = strprintf(_("Error: Could not delete watch only script %s"), HexStr(script));
+            return false;
```
Is `batch.Tx Abort();` required in case any deletion fails here?

```suggestion
            batch.TxnAbort();
            return false;
```

achow101 replied:
No, `TxnAbort` will be called by the destructor.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -1425,6 +1425,12 @@ bool CWallet::IsMine(const CTransaction& tx) const
     for (const CTxOut& txout : tx.vout)
         if (IsMine(txout))
             return true;
+    for (const CTxIn& txin : tx.vin) {
+        const CWalletTx* wtx = GetWalletTx(txin.prevout.hash);
+        if (wtx && wtx->tx->vout.size() > txin.prevout.n && IsMine(wtx->tx->vout[txin.prevout.n])) {
+            return true;
+        }
+    }
```
The purpose here is to identify the transactions whose inputs belong to the wallet ?

achow101 replied:
Yes

furszy replied:
This `IsMine` change will have implications inside `AddToWalletIfInvolvingMe` as well.
We are checking tx ownership using the following conditions: `if (fExisted || IsMine(tx) || IsFromMe(tx))`

So, with this changes, the middle `IsMine` will check what `IsFromMe` currently checks. So.. either here or in a follow-up PR, can remove the `IsFromMe` function entirely.

Other point is that you can simplify the code:
```c++
for (const CTxIn& txin : tx.vin) {
    if (IsMine(txin.prevout)) return true; 
}
```

achow101 replied:
I've removed this change and switched to using `IsFromMe` in addition to `IsMine`. Generally we should avoid changing `IsMine` semantics when possible.

Sjors commented:
I noticed that the migrated `combo()` descriptor and the `addr()` descriptor in the watch-only part of the migrated wallet uses timestamp 0. Can it use the original timestamps instead?

The new bech32m descriptor can't be used in the GUI unless you reload the wallet, since the dropdown is not updated.

Other than that it all works with my test wallet. Will look at the code.

Sjors commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -242,6 +242,9 @@ class ScriptPubKeyMan
     virtual uint256 GetID() const { return uint256(); }
 
+    /** Returns a vector of all the scriptPubKeys that this ScriptPubKeyMan watches */
+    virtual const std::unordered_set<CScript, SaltedSipHasher> GetScriptPubKeys() const { return {}; };
```
220fe61f70f6093a695fd8b6e38e928b63c2b6ad: I assume the goal here is to get faster lookups and random access deletions? E.g. in the `erase()` call in the next commit. 

Afaik there's no guarantee in the standard library that `std::unordered_set` will put elements in the same order between runs. So if deterministic behaviour is the goal of this change, it might be better to just stick to `std::vector` but call `sort()` on it before returning. But that's slower.

Code comment nit: not a "vector"

Suggest renaming commit to: "scriptpubkeyman: Implement GetScriptPubKeys for Legacy" (the current title suggests it's an entirely new method)

achow101 replied:
We don't need deterministic behavior. That's why I elected to use `std::unordered_set` and `SaltedSipHasher` (which is initialized with random salts).

achow101 replied:
Fixed the comment and commit message.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3633,7 +3633,7 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
 
         CTxDestination dest;
-        if (!internal && ExtractDestination(script_pub_keys.at(0), dest)) {
+        if (!internal && ExtractDestination(*script_pub_keys.begin(), dest)) {
```
220fe61f70f6093a695fd8b6e38e928b63c2b6ad: shouldn't this be a loop over all elements anyway, for `combo()`. It seems it would always label the `P2PK` variant before (which `ExtractDestination` turns into `P2PKH`), and a random one after.

Sjors replied:
That broke `tool_wallet.py —descriptors` because `Address Book: ` needs to be 3 now instead of 1.

achow101 replied:
Extracted to a separate commit and fixed the test.

w0xlt commented on src/wallet/wallet.cpp:
```diff
@@ -3946,7 +3946,7 @@ bool DoMigration(CWallet& wallet, WalletContext& context, bilingual_str& error,
             DatabaseStatus status;
             std::vector<bilingual_str> warnings;
-            std::string wallet_name = wallet.Getname() + "_watchonly";
+            std::string wallet_name = wallet.GetName() + "_watchonly";
```
Perhaps the change in `src/wallet/wallet.cpp` in https://github.com/bitcoin/bitcoin/commit/7ab26895d879e8cadd70109bd84c0b7e586c31a2 ('Add migratewallet RPC') could be squashed into https://github.com/bitcoin/bitcoin/commit/211c12bcd4545aa457a2efda17dfbf6a486407c3 ('Implement MigrateLegacyToDescriptor').

Currently, https://github.com/bitcoin/bitcoin/commit/211c12bcd4545aa457a2efda17dfbf6a486407c3 does not compile..

achow101 replied:
Done

achow101 commented:
> The new bech32m descriptor can't be used in the GUI unless you reload the wallet, since the dropdown is not updated.

This appears to be an issue in general. Importing a `tr` descriptor to a wallet that does not have one yet does not cause the dropdown to be updated.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+            // Make the DescriptorScriptPubKeyMan and get the scriptPubKeys
+            auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc));
+            desc_spk_man->AddDescriptorKey(master_key.key, master_key.key.GetPubKey());
+            desc_spk_man->TopUp();
```
fc33d306874ec3bca38c34ff5ac49cbcfa7d832f: I think you need to pass `chain_counter` here, because `DescriptorScriptPubKeyMan::TopUp` defaults to `-keypool`, not to the range you passed into the descriptor.

achow101 replied:
No, it will account for the range_end.

Sjors replied:
Oh wait, I only looked at the first few lines where it sets `target_size`. But then later on it also looks at `m_wallet_descriptor.range_end`

achow101 committed:
Implement LegacyScriptPubKeyMan::DeleteRecords

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+    }
+
+    // Make sure that we have accounted for all scriptPubKeys
+    assert(spks.size() == 0);
```
fc33d306874ec3bca38c34ff5ac49cbcfa7d832f: we should probably fail more gracefully.

achow101 replied:
An assert is used here because it is a critical bug. The purpose is to have something that obviously fails during testing.

Sjors replied:
But can we really rule out that this happens in production on some obscure old weird wallet? If not, it would be better to have the RPC throw and maybe provide some info about the scripts that are left.

achow101 replied:
I'm pretty confident that we can.

Sjors commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -511,6 +513,9 @@ class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProv
     std::set<CKeyID> GetKeys() const override;
     const std::unordered_set<CScript, SaltedSipHasher> GetScriptPubKeys() const override;
+
+    /** Get the DescriptScriptPubKeyMans that have the same scriptPubKeys as this LegacyScriptPubKeyMan */
```
fc33d306874ec3bca38c34ff5ac49cbcfa7d832f  Description could be slightly more elaborate. Maybe mention that it's non-invasive and requires the wallet to be unlocked.

achow101 replied:
Done

Sjors reviewed (commented):
Reviewed fc33d306874ec3bca38c34ff5ac49cbcfa7d832f which looks mostly good. I skimmed over the multisig part though. Will review the other commits later.

achow101 commented:
> I noticed that the migrated `combo()` descriptor and the `addr()` descriptor in the watch-only part of the migrated wallet uses timestamp 0. Can it use the original timestamps instead?

Done



> A simple restart does the trick there. In any case, new funds are safu.

I've changed it to reload the wallet after migrating so that all of the signals and gui stuff get refreshed.

This happened to expose a bug where not all legacy records were being removed from all wallets. I've pushed a change to how we do the deletions to make sure there is no lingering data.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+        // Remove the scriptPubKeys from our current set
+        for (const CScript& spk : desc_spks) {
+            size_t erased = spks.erase(spk);
+            assert(erased == 1);
```
No. This is a check that it erased something rather than it didn't erase more than one. `spks` is a set so it can't erase more than 1 anyways.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+    for (const CHDChain& chain : chains) {
+        for (int i = 0; i < 2; ++i) {
+            // Skip if doing internal chain and split chain is not supported
+            if (chain.seed_id.IsNull() || (i == 1 && !m_storage.CanSupportFeature(FEATURE_HD_SPLIT))) {
```
Generally it can't be.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+            std::string error;
+            std::unique_ptr<Descriptor> desc = Parse(desc_str, keys, error, false);
+            uint32_t chain_counter = std::max((i == 1 ? chain.nInternalChainCounter : chain.nExternalChainCounter), (uint32_t)0);
+            WalletDescriptor w_desc(std::move(desc), 0, 0, chain_counter, 0);
```
No. CHDChains do not have any time metadata.

Sjors replied:
Sad. And their keys don't either?

Sjors replied:
Sad, not even when they were first put in the keypool? Though perhaps that's too risky if we're upgrading a wallet that was itself restored from a backup. The downside of setting 0 is that if a user were to dump the descriptors and restore them into a fresh wallet, they'd have to rescan from genesis. Though that's not the  backup / recovery flow atm.

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+            desc->Expand(0, provider, desc_spks, provider);
+        } else {
+            // Make the DescriptorScriptPubKeyMan and get the scriptPubKeys
+            WalletDescriptor w_desc(std::move(desc), 0, 0, 0, 0);
```
Done

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1711,242 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+        // Get the key origin
+        // Maybe this doesn't matter because floating keys here shouldn't have origins
+        KeyOriginInfo info;
+        bool has_info = GetKeyOrigin(keyid, info);
```
Done.

achow101 committed:
Apply label to all scriptPubKeys of imported combo()

w0xlt commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,407 @@
+            assert_equal(file_magic, b'SQLite format 3\x00')
+        assert_equal(self.nodes[0].get_wallet_rpc(wallet_name).getwalletinfo()["format"], "sqlite")
+
+    def create_wallet(self, wallet_name):
```
nit: makes the purpose of the function clearer.

```suggestion
    def create_legacy_wallet(self, wallet_name):
```

achow101 replied:
Done

bangz55 replied:
Benito Calistro Lopez

tusher456 commented:
32sMEMSwnLLcuxzHptTDksk9auSPoccWiJ

On Fri, 26 Aug 2022, 3:55 am w0xlt, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/functional/wallet_migration.py
> <https://github.com/bitcoin/bitcoin/pull/19602#discussion_r955452249>:
>
> > +        self.extra_args = [[]]
>
> +        self.supports_cli = False
>
> +
>
> +    def skip_test_if_missing_module(self):
>
> +        self.skip_if_no_wallet()
>
> +        self.skip_if_no_sqlite()
>
> +        self.skip_if_no_bdb()
>
> +
>
> +    def assert_is_sqlite(self, wallet_name):
>
> +        wallet_file_path = os.path.join(self.nodes[0].datadir, "regtest/wallets", wallet_name, self.wallet_data_filename)
>
> +        with open(wallet_file_path, 'rb') as f:
>
> +            file_magic = f.read(16)
>
> +            assert_equal(file_magic, b'SQLite format 3\x00')
>
> +        assert_equal(self.nodes[0].get_wallet_rpc(wallet_name).getwalletinfo()["format"], "sqlite")
>
> +
>
> +    def create_wallet(self, wallet_name):
>
>
> nit: makes the purpose of the function clearer.
> ⬇️ Suggested change
>
> -    def create_wallet(self, wallet_name):
>
> +    def create_legacy_wallet(self, wallet_name):
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/19602#pullrequestreview-1086153023>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2RJYJVGIJW4N2AUACNSBDDV27TWZANCNFSM4PJHLFTQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

w0xlt commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,407 @@
+
+        # Contrived case where all the multisig keys are in a single wallet
+        self.log.info("Test migration of a wallet with all keys for a multisig")
+        self.nodes[0].createwallet(wallet_name="multisig0")
+        multisig0 = self.nodes[0].get_wallet_rpc("multisig0")
+        assert_equal(multisig0.getwalletinfo()["descriptors"], False)
```
nit: (or `create_legacy_wallet`, as suggested above).

```suggestion
        multisig0 = self.create_wallet("multisig0")
```

achow101 replied:
Done

w0xlt commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,40 @@ RPCHelpMan simulaterawtransaction()
+    if (!MigrateLegacyToDescriptor(std::move(wallet), context, error)) {
+        throw JSONRPCError(RPC_WALLET_ERROR, error.original);
+    }
+    return NullUniValue;
```
Maybe it can return a success message and others indicating the backup, watchonly and multisig file path?

achow101 replied:
Done

Sjors commented on src/wallet/walletdb.cpp:
```diff
@@ -59,6 +59,7 @@ const std::string WALLETDESCRIPTORCKEY{"walletdescriptorckey"};
 const std::string WALLETDESCRIPTORKEY{"walletdescriptorkey"};
 const std::string WATCHMETA{"watchmeta"};
 const std::string WATCHS{"watchs"};
+const std::unordered_set<std::string> LEGACY_TYPES{CRYPTED_KEY, CSCRIPT, DEFAULTKEY, HDCHAIN, KEYMETA, KEY, OLD_KEY, POOL, WATCHMETA, WATCHS};
```
c50135edb6414fa5fd598f65e13eaa0e639573b9: maybe add comment: `// These keys are deleted by the migratewallet RPC` to ensure we don't add more stuff to this list without making sure they get migrated.

Note to other reviewers (plus some open questions), these types are captured by the backup as follows:
* `CRYPTED_KEY`: when `walletdb.cpp` loads them, it puts them into `mapCryptedKeys` via `LoadCryptedKey()`. `MigrateToDescriptor()` iterates over that map.
* `CSCRIPT`: when `walletdb.cpp` loads them, it puts them into `mapScripts` via `LoadCScript()`. 
   * `MigrateToDescriptor()` iterates over any multisig scripts in that map
   * TODO: I'm unclear what happens to `mapScripts` entries that are not multisig (see comment on earlier commit)
   * caveat: old records with `redeemScript.size() > MAX_SCRIPT_ELEMENT_SIZE` are lost in the upgrade (they're in the backup)
* `DEFAULTKEY`: this is only used to detect corruption, and is not migrated, newer wallets don't have this key
* `HDCHAIN`: `m_hd_chain` which we migrate
* `KEYMETA`: ends up in `mapKeyMetadata` via `LoadKeyMetadata`, which we migrate (at least for every key we have).
* `KEY`:  when `walletdb.cpp` loads them, it puts them into `mapKeys` via `LoadKey()`. `MigrateToDescriptor()` iterates over that map.
* `OLD_KEY`: these can only be loaded with v0.18 and older. They are skipped during wallet load, so they are lost in the migration.
* `POOL`: these are ignored, we regenerate the keypool using descriptors
* `WATCHMETA`: put in `m_script_metadata` by  `LoadScriptMetadata`, and indexed by their `CScriptID`, TODO: same question as with `CSCRIPT`
* `WATCHS` put in `mapWatchKeys`   via `LoadWatchOnly`, and its keys are used to monitor PKH, etc via `ImplicitlyLearnRelatedKeyScripts`. We don't iterate over `mapWatchKeys` directly, so I'm a bit confused how these are migrated.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1711,6 +1712,258 @@ const std::unordered_set<CScript, SaltedSipHasher> LegacyScriptPubKeyMan::GetScr
+
+        std::vector<std::vector<unsigned char>> sols;
+        TxoutType type = Solver(script, sols);
+        if (type == TxoutType::MULTISIG) {
```
35f428fae68ad974abdce0fa905148f620a9443c  What happens to `mapScripts` entries that are not `MULTISIG`?

achow101 replied:
They will be part of `spks` and added to the watchonly wallet. It is not possible to import other scripts without also watching them (and their P2(W)SH). Multisigs are special because we have `addmultisigaddress` which add the multisig but does not watch them. All of the import rpcs will watch.

Sjors commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1964,6 +1964,13 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
     return out;
 }
 
+bool LegacyScriptPubKeyMan::DeleteRecords()
```
c50135edb6414fa5fd598f65e13eaa0e639573b9 : `DeleteLegacyRecords` (since we keep plenty of other stuff)

achow101 replied:
We delete all of the `LegacyScriptPubKeyMan` records (which this function is a member of). There are no remaining LegacySPKM records after this.

Sjors reviewed (commented):
Thanks for the changes. Reviewed another commit, will continue later.

Sjors reviewed (commented):
Reviewed up to a6c0e2f844f589df4f2d19e38c2e350952e1b4bf. Mostly happy. I'm a bit confused about how legacy wallets deal with (unwatched) solvable scripts, so I'll review that aspect again once I wrap my head around it.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+            DatabaseStatus status;
+            std::vector<bilingual_str> warnings;
+            std::string wallet_name = wallet.GetName() + "_watchonly";
+            data->watchonly_wallet = CreateWallet(context, wallet_name, std::nullopt, options, status, error, warnings);
```
I don't think that's any clearer, and would like to keep the function signature small.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+    opts.require_create = true;
+    opts.require_format = DatabaseFormat::SQLITE;
+    DatabaseStatus db_status;
+    std::unique_ptr<WalletDatabase> new_db = MakeDatabase(db_dir, opts, db_status, error);
```
Perhaps for a followup.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+    opts.require_format = DatabaseFormat::SQLITE;
+    DatabaseStatus db_status;
+    std::unique_ptr<WalletDatabase> new_db = MakeDatabase(db_dir, opts, db_status, error);
+    assert(new_db); // This is to prevent doing anything further with this wallet. The original file was deleted, but a backup exists.
```
No. it just contains more granular errors. There's no "success by with caveats" that would be relevant here.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+    }
+
+    std::optional<MigrationData> res = legacy_spkm->MigrateToDescriptor();
+    if (res == std::nullopt) {
```
I don't understand what you're talking about.

Sjors replied:
You're assuming here that the reason `std::nullopt` is returned is that the wallet is locked, rather than other failure.

furszy replied:
The RPC handler is ensuring that the wallet is unlocked (third line there). This check is redundant for now, and might be useful when a migration button is implemented on the GUI.

achow101 replied:
I don't think the message implies that the only reason it would return `nullopt` is because it is locked. However, it is likely that a failure would be due to being locked, hence the message. We do this in a couple of other places too where failure is often being locked, but could be some weird db condition that actually happened.

achow101 committed:
Implement MigrateToSQLite

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+        context.wallets.erase(i);
+    }
+    // Now do the database stuff
+    if (!wallet->MigrateToSQLite(error)) return false;
```
Since these are filesystem errors, I don't think restoring the backup would actually work. Furthermore, because `m_database` has been swapped out, and potentially doesn't have a backing database, I don't think it's safe to continue execution in that state as the `CWallet` would still exist, but nowhere to write data to.

achow101 commented on src/wallet/wallet.h:
```diff
@@ -920,6 +920,16 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     ScriptPubKeyMan* AddWalletDescriptor(WalletDescriptor& desc, const FlatSigningProvider& signing_provider, const std::string& label, bool internal) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+
+    //! Migrate a BDB wallet to SQLite wallet
+    bool MigrateToSQLite(bilingual_str& error) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,454 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+
+    // Get all of the records for DB type migration
+    std::unique_ptr<DatabaseBatch> batch = m_database->MakeBatch();
+    std::map<SerializeData, SerializeData> records;
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -920,6 +920,16 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
+    std::optional<MigrationData> GetDescriptorsForLegacy(bilingual_str& error) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+
+    //! Adds the ScriptPubKeyMans given in MigrationData, removes LegacyScriptPubKeyMan, and moves tx and address book
+    //! to their relevant wallets.
```
Done

w0xlt reviewed (approved):
ACK 3299574f5e27bdb6c12fffd541e7cbfd148ab883

Sjors commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,405 @@
+
+        addr = ms_info["address"]
+        txid = default.sendtoaddress(addr, 10)
+        multisig1.importaddress(addr)
```
fb5e0957e066bd0825947a399242242f1f1a2ae6: why do you need to call `importaddress` after `addmultisigaddress`? Is that makes it marked as `watch-only` (as oppose to the second address)? Why?

achow101 replied:
So that this particular address is being watched and will end up in the `_watchonly` wallet.

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,59 @@ RPCHelpMan simulaterawtransaction()
+    if (res->solvables_wallet) {
+        r.pushKV("solvables_name", res->solvables_wallet->GetName());
+    }
+    r.pushKV("backup_path", res->backup_path);
```
```suggestion
    r.pushKV("backup_path", res->backup_path.u8string());
```

C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\wallet\rpc\wallet.cpp(753,1): message : Reason: cannot convert from 'fs::path' to 'const UniValue' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,59 @@ RPCHelpMan simulaterawtransaction()
+static RPCHelpMan migratewallet()
+{
+    return RPCHelpMan{"migratewallet",
+        "\nEXPERIMENTAL warning: This call may not work as expected and may be changed in future releases\n"
```
nit: No need for the `\n`.
```suggestion
        "EXPERIMENTAL warning: This call may not work as expected and may be changed in future releases\n"
```

achow101 replied:
Done

Sjors commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,405 @@
+        self.generate(self.nodes[0], 1)
+        multisig1.gettransaction(txid)
+        assert_equal(multisig1.getbalances()["watchonly"]["trusted"], 10)
+        assert_equal(multisig1.getaddressinfo(ms_info2["address"])["ismine"], False)
```
fb5e0957e066bd0825947a399242242f1f1a2ae6 : Maybe do `addr2 = ms_info2["address"]` and rename `addr` to `addr1` for clarity.

achow101 replied:
Done

MarcoFalke commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,59 @@ RPCHelpMan simulaterawtransaction()
+            HelpExampleCli("migratewallet", "")
+            + HelpExampleRpc("migratewallet", "")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    std::shared_ptr<CWallet> wallet = GetWalletForJSONRPCRequest(request);
+    if (!wallet) return NullUniValue;
```
nit: For new code, can use correct spacing?

```suggestion
        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
        {
            std::shared_ptr<CWallet> wallet = GetWalletForJSONRPCRequest(request);
            if (!wallet) return NullUniValue;
```

achow101 replied:
Done

maflcko reviewed (commented):
I think it is expected that you run into deadlocks, given that the thread sanitizer doesn't pass CI. 

Also, the Windows compilation fails, see u8string suggestion.

Sjors commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,405 @@
+        multisig1.migratewallet()
+        assert_equal(multisig1.getwalletinfo()["descriptors"], True)
+        self.assert_is_sqlite("multisig1")
+        assert_equal(multisig1.getaddressinfo(ms_info["address"])["ismine"], False)
```
fb5e0957e066bd0825947a399242242f1f1a2ae6 : Maybe use `addr` instead of `ms_info["address"]`

achow101 replied:
Done

Sjors commented on test/functional/wallet_migration.py:
```diff
@@ -0,0 +1,405 @@
+        self.assert_is_sqlite("multisig1_watchonly")
+        assert_equal(ms1_watchonly.getaddressinfo(ms_info["address"])["ismine"], True)
+        assert_equal(ms1_watchonly.getaddressinfo(ms_info["address"])["solvable"], True)
+        assert_equal(ms1_watchonly.getaddressinfo(ms_info2["address"])["ismine"], False)
```
fb5e0957e066bd0825947a399242242f1f1a2ae6: maybe comment here that, because it wasn't watch-only, it doesn't get moved to the watch-only wallet, but to the solvables wallet (as shown below in the test)

achow101 replied:
Done

Sjors reviewed (commented):
As you said:

> Multisigs are special because we have `addmultisigaddress` which add the multisig but does not watch them.

The test could use some comments around that behavior. It correctly demonstrates it though.

Sjors commented:
I was confused about the difference between solvable and watch-only.

* solvable: anything we have _any_ (public) keys for _and_ we know who to spend (e.g. a multisig for which we have one public key, and we know the other two because we have the full redeemscript)
* watch-only: may or may not be solvable, e.g. an unsolvable multisig is one for which we only know the address (scriptPubKey), not the full script, i.e. we don't know which keys it needs

A solvable address is not automatically watch-only, but all import methods except `importmulti` ensure it _is_.

`importmulti`  adds a `redeemscript` so it's solvable, but it's _not_ marked watch-only. The net-effect of that is that we see the full script with `getaddressinfo`, the wallet ignores transactions to it, but we can sign them. This may be useful if you're a co-signer and you don't want those transactions to show up.

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -844,6 +897,7 @@ Span<const CRPCCommand> GetWalletRPCCommands()
         {"wallet", &walletpassphrase},
         {"wallet", &walletpassphrasechange},
         {"wallet", &walletprocesspsbt},
+        {"wallet", &migratewallet},
```
Done

achow101 commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -701,6 +701,59 @@ RPCHelpMan simulaterawtransaction()
+            {
+                {RPCResult::Type::STR, "wallet_name", "The name of the primary migrated wallet"},
+                {RPCResult::Type::STR, "watchonly_name", /*optional=*/true, "The name of the migrated wallet containing the watchonly scripts"},
+                {RPCResult::Type::STR, "solvables_name", /*optional=*/true, "The name of th migrated wallet containing solvable but not watched scripts"},
```
Done

Sjors commented:
The problem is that the solvables wallet is completely useless. It can't sign things the legacy wallet could sign. You might as well put these things in the watch-only wallet, unless I'm missing something.


Another approach could be to handle the special (and only?) case of a multisig wallet more gracefully. The best substitute we can make for how that worked before, is to create a blank descriptor wallet and then import the descriptor with the private keys that we have.

achow101 commented:
> The problem is that the solvables wallet is completely useless. It can't sign things the legacy wallet could sign. You might as well put these things in the watch-only wallet, unless I'm missing something.

The solvables wallet can update a PSBT with the UTXO, scripts, and pubkeys, for inputs that spend the any of the scripts it contains. Because the user had not watched those scripts when the wallet is a legacy wallet, it does not make sense to have them be in the watchonly wallet.

Sjors commented:
The slowness seems to be macOS thing, or even particular to _one_ of my machines.  The exact same wallet takes ~9 minutes to upgrade on one on my macOS machines, and less than a minute on the other and on Ubuntu. When the upgrade is fast, the labels are preserved fine too.

---

It would be useful to know how people use(d) legacy multisig wallets, especially before PSBT. It's not clear to me with which wallet they would craft the transaction.

A solveable-only legacy wallet can sign them and add more metadata, but because it lacks transaction history, they can't craft them. Presumably the user would have different wallet for that purpose, which they may or may not be upgrading separately. As long as that other wallet can produce a PSBT, then our upgraded main wallet can sign it just fine. The _solvable wallet's only purpose then is to backup the redeemscript. They should never need it.

If their other wallet can't produce a psbt, then the upgrade process breaks their ability to sign, since the private keys are now no longer in the same wallet as the redeemscript. So we should warn about that. And we could a utility that takes a raw unsigned transaction and transform it into a PSBT to address that potential problem.

Knowing these use cases would also help us write upgrade documentation for them, to explain what new workflows they should consider.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3654,4 +3654,74 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+            assert(false); // This is a critical error, the new db could not be written to. The original db exists as a backup, but we should not continue execution.
+        }
+    }
+    batch->TxnCommit();
+    return true;
```
`TxnCommit` can fail and return false (same as `TxnBegin`).

Same as it's done above, can throw a critical error.

achow101 replied:
Added some asserts for this and `TxnBegin`.

Sjors commented:
tACK 53e7ed075c49f853cc845afc7b2f058cabad0cb0

I think this is good enough for an `experimental` RPC. Hopefully we can figure out why some macOS machines have difficulty with the RPC, but if not, we can just mention that in the release notes. Similarly we can revisit the way solvable multisig is handled; anyone using that feature will know how to restore the backup.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/commit/53e7ed075c49f853cc845afc7b2f058cabad0cb0

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3726,4 +3726,400 @@ bool CWallet::MigrateToSQLite(bilingual_str& error)
+    for (const auto& addr_pair : m_address_book) {
+        // Labels applied to receiving addresses should go based on IsMine
+        if (addr_pair.second.purpose == "receive") {
+            if (!IsMine(addr_pair.first)) {
```
In commit "Implement MigrateLegacyToDescriptor" (0bf7b38bff422e7413bcd3dc0abe2568dd918ddc)

It seems like this should say `IsMine` not `!IsMine`. Also I'm not sure why `addr_pair.second.purpose == "receive"` check is necessary. It seems like just checking IsMine should be enough, and it would be less robust to rely on purpose field being present.

furszy replied:
It's because the migration process divides data between different wallets.
It first deletes and unloads the legacy spkm from the main wallet (check the beginning of the function), and setup new descriptors. Then moves the data to two possible wallets; a solvable and a watch-only wallet.

So the `!IsMine` means that the record in the addressbook is from the legacy spkm (not loaded anymore), so it needs migration to the new wallets.

It also took me a while to get it while was rewriting it for #26836.
